### ISSUE TITLE
Netflix error with 0.13.4 & 0.13.5 but not 0.13.2

### ISSUE THREAD

**Issue by knappster1** [asciidisco/plugin.video.netflix#362]
*I'm submitting a ...*
  - [x] bug report
  - [ ] feature request
  - [ ] support request

## General information
I am running the Netflix plugin in Kodi Alpha 1 that was built from source for my 64-bit Arch based linux OS.  

### Prerequisites

* [x] Are you running the latest version?

### Description

The Netflix plugin 0.13.2 still works fine, but when it automatically upgraded to 0.13.4 I would get an error on startup of Kodi and when I try to launch the Netflix plugin.

"Netflix Error
Check the log for more information"

### Steps to Reproduce

1. Launch Kodi (error message is displayed as plugins are loaded)
2. Select Netflix plugin (error message is displayed again)

**Expected behavior:**
1. Launch Kodi (no error messages)
2. Select Netflix plugin (Netflix launches and runs as it did in 0.13.2)

**Actual behavior:**
Explained in steps to reproduce.

### Installation

* [ ] I installed the plugin via zip from the Releases page
* [x] I´m using the Netflix Repo
* [ ] I´m using a different source (Please tell which)

### Operating System

* [x] Linux (x86/x64)
* [ ] OSX (x86/x64)
* [ ] Windows (x86/x64)
* [ ] Linux (ARM)
* [ ] Android

#### Additional informatin on the environment

[Descripe your environment a bit more detailed, are you using LibreElec f.e.]

## Debug log

Here is the pastebin log snippet for when I first launch Kodi:
https://pastebin.com/ZZE5qyNe

Here is the pastebin log snippet for when I try to launch the Netflix Plugin:
https://pastebin.com/WyKDYyrR

## Other information
This looks very similar to the issue reported here:
https://github.com/asciidisco/plugin.video.netflix/issues/357

For him, the upgrade from 0.13.4 to 0.13.5 apparently fixed it, but not so much for me.

**Comment by Dodobibi** [comment]
Excuse my english, I'm french...

I have the same problem.

It's because falkorCache has uncorrect escape string, or escape string is not encoded or double encoded, I don't know...

It's not general, it depends of the titles of the videos in falkorCache (different by users), and if these titles contain double-quote. So, falkorCache json was wrong... and can't be parsed !

I use this workaround : replace `\"` by `\\"` in content before `encode('utf-8').decode('string_escape')`

replace
`content = content.encode('utf-8').decode('string_escape')`
by
`content = content.replace('\"', '\\"').encode('utf-8').decode('string_escape')`
at [https://github.com/asciidisco/plugin.video.netflix/blob/master/resources/lib/NetflixSession.py#L143](url)

This allow to conserve the escape string (for double-quote in string) in json objects after decode('string_escape'), and all working nice now (for me)

**Comment by knappster1** [comment]
@Dodobibi I made the change you suggested, but it did not make any difference in my case.  If I get an opportunity, I will do some further investigation this weekend.  My guess is in the parentheses in the title, but that is purely speculation since that is the only difference I can see with that title.  I will report back any findings.

**Comment by Dodobibi** [comment]
Are you tried to restart kodi after change ?
can your post your complete falkorCache (WITHOUT sensible informations in profiles)
I'm pretty sure it's a wrong json parse.

**Comment by dgktkr** [comment]
> It seems that:
https://github.com/asciidisco/plugin.video.netflix/blob/master/resources/lib/NetflixSession.py#L176
is corrupting the json data somehow (for me anyway)

I've wondered what the pattern `r"\{([^)]+)\}"` in NetflixSession.py line 176 is intended to do. After having studied regular expressions a bit over the years (i.e. I'm a noobie), my understanding is that that pattern captures all the characters in a string between the outer curly brackets unless there is a ')' somewhere in there. In that case, the search fails and findall() returns an empty list, which would result in get_profiles() returning an empty dictionary. In other words, if Netflix includes a ')' anywhere in falkorCache, the add-on will extract no useful profile from it.

If the intent of the search pattern is to capture the falkorCache JSON object, perhaps it might be better to do the search with `search(r"netflix.falkorCache\s*=\s*(.*?);</script>",content)` with a few other changes to support that search.

**Comment by knappster1** [comment]
@dgktkr I was trying to figure the `r"\{([^)]+)\}"` expression out as well, but it was late last night and I was just hacking away at it.  I originally thought the issue was with the single quotes and I was able to revert those changes, but when I got to the information that I posted I realized that there was more going on and I would need to sit down this weekend and study the expression to figure out what it is actually doing (unless the developers have addressed it before I get to it)

**Comment by Dodobibi** [comment]
This regular expression is for extract json from text  (bad) extracted from html, and add { and } for create a textual json object.
I think is not a good solution, but...
The problem is quote (single and double). The decode encode (line 143) escape the quotes in string and this produce a wrong json text...
Look at falkorCache juste before line 176.
I will do test this weekend and I will propose a PR with a more clean solution for extracting falkorCache from content.

Have a nice weekend !

**Comment by knappster1** [comment]
@Dodobibi falkorCache just before line 176 seems fine (with regards to JSON syntax) apart from the extraneous text at the end.  For my account, the problem is occurring with line 176 where text is disappearing (and breaking JSON syntax).  It's possible that it is related to single and double quotes somehow, but it seems that regex is ultimately what is stripping the content out and breaking JSON syntax.  I can't confirm that the JSON syntax is the only issue, though.

As a sidenote, I was looking at the falkorCache on the web browser on another machine and the content appeared to be completely different and did not contain this title with parentheses in it.

**Comment by Dodobibi** [comment]
It seems fine, but is not.
Try to clean the end, and past it in an code editor wich support json, you will seen errors, just where the quotes.

**Comment by knappster1** [comment]
@Dodobibi I don't know if it is safe to paste the entire string, but I will show you from jsoneditoronline.org
Before line 176 (cleaned up): 
![before](https://user-images.githubusercontent.com/39322426/40251786-f45575a2-5a9f-11e8-9d2f-dd9cb602f9d3.png)

After line 176:
![after](https://user-images.githubusercontent.com/39322426/40251796-faf8edda-5a9f-11e8-955b-982b2607219d.PNG)

**Comment by Dodobibi** [comment]
@knappster1  How do you debug ?

Sorry to insist, but if you copy/past from console, or log, the quotes are escaped (by python print function), then, the json semms fine, but json really passed to json.parse (L176), is not...

I debug with pydev, I put a breakpoint and extract the json text before line 176 with debugger.
Look at this excerpt (in notifications.fulllist.message.templateJSON.elements.altText) : 
{
  "templateFormat": "TEXT",
  "messageGuid": "2591B14FE51AF0DDD5C84F7115904F4C6452B41D",
  "boxArtSecure": true,
  "uiPlatform": "AKIRA",
  "eventGuid": "604843E4F78AACA09DA9FA5DEBEC35F46452AE84",
  "boxArtFormat": "tcard",
  "templateJSON": {
    "elements": {
      "image": {
        "target": {
          "url": "http://www.netflix.com/title/80198305",
          "type": "DISPLAY"
        },
        "altText": "Chapitre trente-cinq\\u00A0: "Le meilleur des mondes"",
        "urls": {
          "112x63": "https://art-s.nflximg.net/70a76/0d72b291cd6d02bb40c58736c2c37488b4170a76.jpg"
        }
      },
      "text": {
        "target": {
          "url": "http://www.netflix.com/title/80198305",
          "type": "DISPLAY"
        },
        "showTimestamp": true,
        "body": "Un nouvel \\u00E9pisode est disponible",
        "header": "Riverdale"
      }
    },
    "videoType": "show",
    "videoId": 80198305,
    "layout": "IMAGE-TEXT"
  }
}`

Try it in jsoneditoronline.org, and you will see than "Le meilleur des mondes", quotes are not correctly escaped.

It is for this reason that the regex produces a bad json : "DISPLAY"", ""112x63" The comma is escaped... and there are missing elements...

Remove the quotes around "Le meilleur des mondes", or escape their with \, and json will OK.
This is the source of the problem.

Regards,

**Comment by knappster1** [comment]
@Dodobibi I am just debugging from the printout in the debug log from the print command that I've added into the script.  Can you recommend how to setup the debugging with pydev and setting up the breakpoint?  Is there already a guide somewhere?

**Comment by dgktkr** [comment]
@Dodobibi, my problem [https://github.com/asciidisco/plugin.video.netflix/issues/357](url) has returned and your solution of replacing 

`content = content.encode('utf-8').decode('string_escape')`

with

`content = content.replace('\"', '\\"').encode('utf-8').decode('string_escape')` fixes the problem.

I have been able to write to a file the original data received from the Netflix website. There, the quote marks inside quote marks are properly escaped, so the falkorCache JSON is valid. After 

`content = content.encode('utf-8').decode('string_escape')`

is applied, the JSON is no longer valid. 

Your solution fixes that. Thank you.

**Comment by liberty-developer** [comment]
Thanx for your support, guys! Yes, this place is something which makes me headache, too.
For sure the quoting must be fixed somehow, but the regex is not safe here, I agree.

Unfortunately json parser returns nothing if I pass more data as expected, my first approach was let json parser findd the end of the block, but I was without success. Maybe someone find a solution to let JSON parser parse until an error occurs.

Could be possible with overriding this and that, but I'm not the py dev.

**Comment by Dodobibi** [comment]
OK,
There is two problems :
- Double escaping the double-quotes for keeping escape after decode unicode to string
- Extracting json object correctly (without pollution at the end)

Can you test this :

replace line 143
`content = content.encode('utf-8').decode('string_escape')`
by
`content = content.replace('\"', '\\"').encode('utf-8').decode('string_escape')`
This resolve the issue when double-quotes are in string.

and at line 173
```
        # find falkor cache
        falkorCache = content[content.find('netflix.falkorCache = '):]
        # Extract JSON
        falkorCache = recompile(r"\{([^)]+)\}").findall(falkorCache)
        if not falkorCache:
            return profiles
        falkorDict = json.loads('{' + falkorCache[0] + '}')
        _profiles = falkorDict['profiles']
```
by
```
        # find falkor cache and extract JSON text
        falkorCache = recompile(r"netflix\.falkorCache\s*=\s*(.*?);\s*</script>").findall(content)
        if not falkorCache:
            return profiles
        falkorDict = json.loads(falkorCache[0])
        _profiles = falkorDict['profiles']
```
This returns a json object without pollution at the end, then it will be parsed by json.loads correctly and completly. This should resolve the other issue (@knappster1 )

No need to add braces around falkorCache[0] : `'{' + falkorCache[0] + '}'` because the regex correctly extract the json object and completly with braces.

and finally at line 145 
`react_context = recompile('reactContext(?s)(.*);').findall(content)`
by 
`react_context = recompile(r"netflix\.reactContext\s*=\s*(.*?);\s*</script>").findall(content)`
This just for extracting reactContext cleanly, in the same way than falkorCache, without pollution.

@dgktkr Thank you for your help with regex. I think it's a good solution.

@knappster1 Can you confirm me if this patch resolve your issue (when brace in title) ?

If this resolve all issues, @liberty-developer : Can you push theses patchs ?

**Comment by liberty-developer** [comment]
Yes sure, will push things tonight if we're all fine with this

**Comment by FloridaStream** [comment]
Sample problem here, windows, some times i can login, but after, get same problem, i cant fix reinstaling kodi or netflix addon :(

**Comment by knappster1** [comment]
@Dodobibi I made the 3 changes to 0.13.5 that you suggested and all seems to work fine for me.

**Comment by Dodobibi** [comment]
@liberty-developer : I cloned the faktor branch and **all seems work fine**.
PS : If you can you merge the #354 too ? (Add French translation, but with one correction : Recommendations = Recommand**a**tions) It would be perfect !

**Comment by liberty-developer** [comment]
done, can we close this issue?

**Comment by liberty-developer** [comment]
thx guys, was a pleasure with you

**Comment by liberty-developer** [comment]
@dgktkr @Dodobibi there seem to be new issues now after the commit:

18:20:42.612 T:139984451397376   ERROR:   File "/var/lib/vdr/.kodi/addons/plugin.video.netflix/resources/lib/NetflixSession.py", line 282, in login
18:20:42.612 T:139984451397376   ERROR:     content=page.text)
18:20:42.612 T:139984451397376   ERROR:   File "/var/lib/vdr/.kodi/addons/plugin.video.netflix/resources/lib/NetflixSession.py", line 150, in extract_inline_netflix_page_data
18:20:42.612 T:139984451397376   ERROR:     '"' + item + '":(.*?)"(.+?)"').findall(react_context[0])

any clue what could be wrong there?

**Comment by Dodobibi** [comment]
OK, I think I know why...

Because double escape the content with `.replace('\"', '\\"')`, then `'"' + item + '":(.*?)"(.+?)"').findall(react_context[0])` does not work,

Two way to solve this :

- adjust regex on line 150 for work with escaped double quotes (more fast)
- or parse reactContext with json.loads and access it with full path of properties of items (I think is more clean but more complicated.

You can try first solution with `'\\"' + item + '\\":(.*?)\\"(.+?)\\"').findall(react_context[0])`, but I'm not sure until I can test it.

I can't test now, but I will propose a workaround more late this evening, and I will test all other regex if there are another in code.

**Comment by liberty-developer** [comment]
Thx! this is really an ugly place, would be interesting if the JSON is so poor already if we load it or if we break it when we decode to / from utf-8 after retrieval

**Comment by dgktkr** [comment]
I don't know if this has any bearing on the current problem, but after staring at the code for a while it dawned on me that lines 149 and 150 of NetflixSession.py (except for the line break) read

`            match = recompile(
                '"' + item + '":(.*?)"(.+?)"').findall(react_context[0])`

Showing that a raw string is not being used for the search pattern. Would it help if that was changed?

`            match = recompile(
                r'"' + item + '":(.*?)"(.+?)"').findall(react_context[0])`

dgktkr

**Comment by knappster1** [comment]
@dgktkr I am confused.  0.13.5 has worked well for me once I made these changes:
https://github.com/asciidisco/plugin.video.netflix/issues/362#issuecomment-390994868

I think that @liberty-developer is talking about a different user.  I have not yet upgraded to 0.13.6

As best I can tell, my issue was related to a video with `)` in the recently watched list.  For example `Beauty and the Beast (2017)`

Others seem to be having issues with other characters as well.  When I upgrade to 0.13.6 I will report if I experience any issues.

**Comment by liberty-developer** [comment]
Yes, its a user form kodinerds forum, so not reproducable here.
The user is happy with 0.13.5, but not with 0.13.6

I still believe that we should not decode the raw incoming data before we json'ed it.
@knappster1 because you can reproduce issues you'll be a good testing candidate :-)

**Comment by dgktkr** [comment]
@liberty-developer It seems to me that issue #362 here is different from the one the most recent one you describe a couple of posts above (the same as #373 and #377?), which occurs earlier in the startup sequence of the add-on (line 150 of NetflixSession.py).

**Comment by Dodobibi** [comment]
Look at PR #378 

Overall, I improved html (encoding) responses to remove json errors. Can solve issue #373, and other related to json.loads.
Be sure we must retrieve html with `response.content` and not with `content.text`. Because json is inlined in html, and html is utf-8 encoded, `response.content` give a raw content, wich we can manipulate and decoding correctly, while `content.text` return an altered escaped json, 

I also improved the method to retrieve reactContext items (more cleanly, and without a regex, possible 
 source of problems).
> parse reactContext with json.loads and access it with full path of properties of items (I think is more clean).

I add a hook `react_context = react_context.replace('^(?!\s*$).+', '')`. It's because when we are unlogged, Netflix send regex in the reactContext json (wich correct in javascript, but not in json), which cause wrong json (Yes, again !!! A workaround consist to remove these regex, not necessary).

I did a lot of test, and hope everything works fine now.

**Comment by liberty-developer** [comment]
wow, cool!! this is greeat. Just wanted to start with an user inverstigating further!
I'll test with the user and come back

**Comment by knappster1** [comment]
@liberty-developer 0.13.6 works for me without modification.
@Dodobibi I downloaded your NetflixSession.py from https://github.com/asciidisco/plugin.video.netflix/pull/378 and copied it over what came with 0.13.6 and it works well for me, too.

**Comment by liberty-developer** [comment]
@Dodobibi  unfortunately still issues with umlaute.
I asked the user to write the content to file, I have it here but don't want to share public.

Can you pls send me an email (the one attached to my github account) so we can share the information ? I'll be unavailable until tonight (CET),

**Comment by Dodobibi** [comment]
> react_context = recompile(r"netflix.reactContext\s*=\s*(.?);\s</script>").findall(content)[0]
IndexError: **list index out of range**

It seems that content has no reactContext ?!?

@liberty-developer Yes, I can try to debug with your files. I send you my mail.

**Comment by liberty-developer** [comment]
@Dodobibi we should do the regexp beore we do the decoding / escaping.
Seems to solve the issues with the data I send you via mail.

- react_context = recompile(blabla
- react:context = react:context.replace('\"', '\\"').decode('unicode_escape')

**Comment by liberty-developer** [comment]
@knappster1 can you apply patches?
patch -p 1 < test.patch

Curious if this brakes anything for you

```
diff --git a/resources/lib/NetflixSession.py b/resources/lib/NetflixSession.py
index c050eec..e2b40f1 100644
--- a/resources/lib/NetflixSession.py
+++ b/resources/lib/NetflixSession.py
@@ -141,10 +141,10 @@ class NetflixSession(object):
         self.nx_common.log(msg='Parsing inline data...')
         items = self.page_items if items is None else items
         user_data = {'gpsModel': 'harris'}
-        content = content.replace('\"', '\\"').decode('unicode_escape')
         # find <script/> tag witch contains the 'reactContext' globals
         react_context = recompile(r"netflix\.reactContext\s*=\s*(.*?);\s*</script>").findall(content)[0]
         # Hook for cleaning json when unloggued
+        react_context = react_context.replace('\"', '\\"').decode('unicode_escape')
         react_context = react_context.replace('^(?!\s*$).+', '')
         react_context_dict = json.loads(react_context, encoding='utf-8')
         # iterate over all wanted item keys & try to fetch them
@@ -181,7 +181,11 @@ class NetflixSession(object):
         falkorCache = recompile(r"netflix\.falkorCache\s*=\s*(.*?);\s*</script>").findall(content)
         if not falkorCache:
             return profiles
-        falkorDict = json.loads(falkorCache[0], encoding='utf-8')
+        falkorCache = falkorCache[0]
+        falkorCache = falkorCache.replace('\"', '\\"').decode('unicode_escape')
+        falkorCache = falkorCache.replace('^(?!\s*$).+', '')
+
+        falkorDict = json.loads(falkorCache, encoding='utf-8')
         _profiles = falkorDict['profiles']

         for guid in _profiles:
```

**Comment by Dodobibi** [comment]
@liberty-developer I only seen your post.

You can do recompile... before replace, decode... but json.loads can return an error, (because possible line-feed in string). It's safe to remove '\\n' before decoding

**Comment by liberty-developer** [comment]
you didn't receive my mail?

**Comment by liberty-developer** [comment]
nevermind, misunderstood your previous post, all fine.

I'll come back to you if I have the original unicode content.

**Comment by dgktkr** [comment]
FWIW The add-on with all the latest patches (https://github.com/asciidisco/plugin.video.netflix/pull/378/commits/c0b2d9672f3d8ae67a3454323d9d55ed21a9b2d8 + the one 4-5 posts above this one) works for me even though I have a recently viewed a movie with an umlaut in the title (3 Türken und ein Baby). 

Before applying the patches and before starting to view that movie, the add-on (0.13.6) had worked well. Then I found that movie and watched for a short time on Firefox. After that the add-on failed to run. But now that I've applied the recent patches, the add-on runs well again.

**Comment by liberty-developer** [comment]
@Dodobibi I believe that \n is ok for JSON parser.
Do you want to add my diff to your PR ?

I'll wait until tonight for the OK of user with the issues.
If things work fine for him, we can merge

**Comment by Dodobibi** [comment]
@liberty-developer \\n is OK, but after decoding, it's a real endline "LF", wich do falling regex and else if reges pass, json.loads fail too.
It's the first reason than your user's file is incorrect. regex fail.
 But I find a clean solution for this (I was sleep on it): 

- Use re.DOTALL flag to regex (so regex is compliant with endline in (.*?) grabbing)
- And use strict=False parameter in json.loads (this authorize json to have endline in string too !)

I do tests again with special characters in strings, braces, endline, etc ... and I will do propose a new fully PR (including your diffs) in one or two hours.
PS : We can do recompile before replace, decode (DOTALL is not necessary in this case, but keeping it, it's safely. and do it first increases perfs because size of string to replace... decode... is shorter)

**Comment by Dodobibi** [comment]
I propose a new PR #380  wich replace #378 with little refactor for regex and json.loads, with comments.

@liberty-developer Can you delete #378 please ? It's obsolete.

My tests successfully passed, but I can't test with umlaut or other special char because I don't know how netflix encode them initially. If users can test it (@knappster1 @dgktkr )? I think they are correctly encoded, so, they must pass. But...

I add a helper to save content at raw to disk (commented by default) L155-156, if this can help to analyse the original content... but **not paste it on Internet**. If uncommented, the file is saved at userdata/addon_data/plugin.video.netflix/**raw_content**

I waiting your returns.

**Comment by knappster1** [comment]
@Dodobibi if https://github.com/asciidisco/plugin.video.netflix/issues/362#issuecomment-392361646 is still the latest, I ran this patch and things look okay for me.  Unfortunately I've just noticed that the title is no longer on my recently watched list, so I cannot be sure that I am still a good test candidate for this.

I am currently unable to add anything to My List through the addon so I will create a separate issue for that.  I added it through a browser just to see if it would lead to any json issues, but so far, so good.

**Comment by Dodobibi** [comment]
@knappster1 It seems that Add and Remove from 'My list' work, but the addon refrech this list only after reloading profil, or restart. (for me)

**Comment by knappster1** [comment]
@dodobibi https://github.com/asciidisco/plugin.video.netflix/pull/380 works well for me.

**Comment by liberty-developer** [comment]
The user I'm supporting has:

```
18:20:14.071 T:140231412016896   DEBUG: [Netflix] Started (Version 0.13.6)
18:20:14.071 T:140231412016896   DEBUG: [Netflix] "Navigation::router" called with arguments :paramstring = :
18:20:14.071 T:140231412016896   DEBUG: [Netflix] "Navigation::before_routing_action" called with arguments :params = {}:
18:20:14.072 T:140231057782528   DEBUG: [Netflix] Loading cookies from memory
18:20:14.072 T:140231057782528   DEBUG: [Netflix] Cookie expires: 1553788939 / 1527524414
18:20:14.074 T:140231412016896   DEBUG: [Netflix] "Navigation::before_routing_action" returned: {}
18:20:14.074 T:140231412016896   DEBUG: [Netflix] "Navigation::show_profiles" called
18:20:14.221 T:140231412016896   DEBUG: [Netflix] "Navigation::show_profiles" returned: None
18:20:14.221 T:140231412016896   DEBUG: [Netflix] "Navigation::router" returned: None
```
So seems that we're still not done ;-( Waiting for the debug file....

**Comment by Dodobibi** [comment]
@liberty-developer Can you give the url of posts on kodinerd where your user talk ? 
I would like to understand his problem.

**Comment by liberty-developer** [comment]
nvm, all fine after clean login

**Comment by Dodobibi** [comment]
Yes 👍

**Comment by Dodobibi** [comment]
This add-on is launched as "service" by Kodi.
It work in background.
When I start Kodi, I must wait 10-20 seconds to this add-on fully initialized before navigate in Netflix, else I have error too, but in my case, it's not a real error, just wait complete init and go to Netflix again.

**Comment by dgktkr** [comment]
Perhaps a minor nit with #380 and 0.13.7. 

Lines 155 and 156 of NetflixSession.py are

     # with open(self.data_path + 'raw_content', "wb") as f:
     #     f.write(content)

which I think won't work if 'content' contains characters that are outside the range of 0-127. (Please someone correct me if that's not right.) Also, will the file be written if the script errors out?

I suggest instead

      # with open(self.data_path + 'raw_content', "wb") as f:
      # f.write(content.decode('utf-8').encode('unicode_escape'))
      #     f.close()

**Comment by Dodobibi** [comment]
The goal is to save the original content received without touch it's encoding. Otherwise it is useless.

It work as this because the file is writing as binary `"wb"`.
